vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/ApolloWSConnectionHandler.java

```diff
@@ -94,22 +101,62 @@ private void handleMessage(JsonObject jsonObject) {
       return;
     }
 
-    ApolloWSMessage message = new ApolloWSMessageImpl(serverWebSocket, type, jsonObject);
+    ApolloWSMessageImpl message = new ApolloWSMessageImpl(serverWebSocket, type, jsonObject);
 
     Handler<ApolloWSMessage> mh = apolloWSHandler.getMessageHandler();
     if (mh != null) {
       mh.handle(message);
     }
 
+    Handler<ApolloWSConnectionInitEvent> connectionInitHandler = apolloWSHandler.getConnectionInitHandler();
+
     switch (type) {
       case CONNECTION_INIT:
-        connect();
+        if (!connectionInitialized.compareAndSet(false, true)) {
+          sendMessage(opId, ERROR, "CONNECTION_INIT can only be sent once");
+          break;
+        }
+        if (connectionInitHandler != null) {
+          connectionInitHandler.handle(new ApolloWSConnectionInitEvent() {
+            @Override
+            public ApolloWSMessage message() {
+              return message;
+            }
+
+            @Override
+            public Promise<Object> init() {
+              return connectionPromise;
+            }
+          });
+        } else {
+          connectionPromise.complete();
+        }
+        connectionPromise.future().onComplete(ar -> {
+          if (ar.succeeded()) {
+            connect();
+          } else {
+            sendMessage(opId, CONNECTION_ERROR, ar.cause().getMessage());
+            context.setTimer(10, timeout -> serverWebSocket.close((short)1011));
+          }
+        });
         break;
       case CONNECTION_TERMINATE:
         serverWebSocket.close();
         break;
       case START:
-        start(message);
+        if (!connectionInitialized.get()) {
+          sendMessage(opId, ERROR, "CONNECTION_INIT has to be sent before START");
```

Comment:
Same question here: terminate the connection?

Reply:
implemented